[build] Split stdlib to it's own package.

ejgallego opened the pull request:
We introduce a new package structure for Coq:

- `coq-core`: Coq's OCaml tools code and plugins
- `coq-stdlib`: Coq's stdlib [.vo files]
- `coq`: meta-package that pulls `coq-{core,stdlib}`

This has several advantages, in particular it allows to install Coq
without the stdlib which is useful in several scenarios, it also open
the door towards a versioning of the stdlib at the package level.

The main user-visible change is that Coq's ML development files now
live in `$lib/coq-core`, for compatibility in the regular build we
install a symlink and support both setups for a while.

Note that plugin developers and even `coq_makefile` should actually
rely on `ocamlfind` to locate Coq's OCaml libs as to be more robust.

There is a transient state where we actually look for both
`$coqlib/plugins` and `$coqlib/../coq-core/plugins` as to support
the non-ocamlfind plus custom variables.

This will be much improved once #13617 is merged (which requires this
PR first), then, we will introduce a `coq.boot` library so finally
`coqdep`, `coqchk`, etc... can share the same path setup code.

IMHO the plan should work fine.

rgrinberg commented:
> However this may break 3rd party applications, as now the library name will be coq-base.kernel etc

Not necessarily. In dune, we already have a `deprecated_library_name` stanza to facilitate moving libraries between packages painlessly.

In any case, this suggestion looks quite good.

LasseBlaauwbroek commented on coq.opam:
```diff
@@ -21,6 +21,7 @@ license: "LGPL-2.1"
 version: "dev"
 
 depends: [
+<<<<<<< HEAD
```
This should probably not be here

ejgallego commented:
Anyone willing to shepherd this, note that #13655 tries to do the same. So IMHO it is worth going ahead.

Zimmi48 commented:
I think a lot of people would be in favor of this split, but it is interesting to consider what will be the most backward-compatible path going forward. I prefer the virtual `coq` package, including an explicit `coq-stdlib` package over the solution proposed by #13655. However, AFAICS, this PR doesn't yet implement such a virtual `coq` package, does it?

Also, is there any chance to use @rgrinberg's proposal to give even more backward-compatibility?

> Not necessarily. In dune, we already have a `deprecated_library_name` stanza to facilitate moving libraries between packages painlessly.

ejgallego commented:
> I think a lot of people would be in favor of this split, but it is interesting to consider what will be the most backward-compatible path going forward. I prefer the virtual `coq` package, including an explicit `coq-stdlib` package over the solution proposed by #13655. However, AFAICS, this PR doesn't yet implement such a virtual `coq` package, does it?

If it doesn't it is an oversight, it should do it. This PR is basically ready except for fixing the CI.

> Also, is there any chance to use @rgrinberg's proposal to give even more backward-compatibility?

Can try, but not critical, given than we are not stable for the ML API so people in 8.14 can just move `coq -> coq-core`.

rgrinberg commented:
> Anyone willing to shepherd this, note that #13655 tries to do the same. So IMHO it is worth going ahead.

Would it make sense to get rid of the makefile based build system and fully transition to dune first? I'm guessing that it will be much more difficult to maintain backwards compat well with both build systems.

ejgallego commented:
> Would it make sense to get rid of the makefile based build system and fully transition to dune first? I'm guessing that it will be much more difficult to maintain backwards compat well with both build systems.

There is #13617 , so indeed we will merge that one first hopefully.

ejgallego commented:
Some minor nits pending, but overall this should be ready for review.

ejgallego commented:
> Some minor nits pending, but overall this should be ready for review, in particular #13617 could make great use of this PR.

I'm happy with the state of this, modulo the TODO in the PR description but IMHO it should work.

Needs a shepherd.

gares commented on clib/cPath.mli:
```diff
@@ -0,0 +1,18 @@
+
+val make : string list -> t
+val choose_existing : t list -> t option
+
+(* plugins_directory *)
+(* share_file : t ->  *)
```
```suggestion
```

ejgallego replied:
Fixed thanks!

gares commented on clib/cPath.mli:
```diff
@@ -0,0 +1,18 @@
+(************************************************************************)
+
+(* To be made opaque one day *)
+type t = string
```
Can't you make it opaque right now? (you are introducing the type, nobody is relying on its definition)

ejgallego replied:
I tried that some time ago, was far from trivial, and I was using dune `Path.t` API as a reference.

So indeed that is a good thing to do, whether to do in this PR or later I don't mind; but would introduce quite a few `Path.from_string`.

ejgallego replied:
Made it `private`, so it is a good start IMHO [and Path.t -> string is not so risky IMHO], it is a bit annoying to use, but hopefully when we consolidate the loadpath handling the current api will be much improved.

gares commented on coq-stdlib.opam:
```diff
@@ -0,0 +1,28 @@
+languages (e.g. the CompCert compiler certification project, or the
+Bedrock verified low-level programming library), the formalization of
+mathematics (e.g. the full formalization of the Feit-Thompson theorem
+or homotopy type theory) and teaching.
```
This description needs to be updated.

ejgallego replied:
Updated, thanks.

gares commented on coq-core.opam:
```diff
@@ -0,0 +1,34 @@
+languages (e.g. the CompCert compiler certification project, or the
+Bedrock verified low-level programming library), the formalization of
+mathematics (e.g. the full formalization of the Feit-Thompson theorem
+or homotopy type theory) and teaching.
```
This description needs to be updated.

Zimmi48 replied:
Indeed, this description should make it clear that this is an internal package not fit for direct installation by most users.

ejgallego replied:
Updated, let me know what you think.

Zimmi48 replied:
LGTM.

gares commented on test-suite/misc/printers.sh:
```diff
@@ -1,7 +1,11 @@
 #!/bin/sh
 
+set -x
+
```
```suggestion
```

ejgallego replied:
Thanks, done.

gares commented on test-suite/misc/printers.sh:
```diff
@@ -1,7 +1,11 @@
+
 command -v "${BIN}coqtop.byte" || { echo "Missing coqtop.byte"; exit 1; }
 
+echo "XXXXX $INSIDE_DUNE"
+
```
```suggestion
```

gares reviewed (commented):
If I install just `coq-core` does `coqtop` work? where is the prelude? Shouldn't coq-core depend on coq-stdlib?

ejgallego commented:
> If I install just `coq-core` does `coqtop` work? where is the prelude? Shouldn't coq-core depend on coq-stdlib?

It works if you pass `-noinit`, for example coq-core + hott is fully functional.

ejgallego commented:
Thanks for pointing the leftovers / debug code, will fix after the CI run.

gares commented:
> It works if you pass `-noinit`, for example coq-core + hott is fully functional.

Then I think it should be stated very clearly in the package description. I find it very weird that a package provides a binary that does not work unless you specify a very technical option. IMO you should make the error message from `coqlib()` mention that as well.

ejgallego commented:
> Then I think it should be stated very clearly in the package description. I find it very weird that a package provides a binary that does not work unless you specify a very technical option. IMO you should make the error message from `coqlib()` mention that as well.
> 
> The alternative is to have a virtual package (in the debian sense) but I don't know if opam has it. I mean something like `coq-prelude` which is both provided by `coq-stdlib` but also `hott-stdlib` so that you get at least one of the two installed and hence your binary works.

Note that we already have the `coq` package, so I'd expect only expert users to install `coq-core` only, not sure how verbose about such low-level options we should be.

opam can do it's own virtual-style packages using or, but I am also not sure we would benefit from another package [whereas the split seems useful as it has been suggested independently by 4 or 5 people already]

Anyways I'm open to any kind of setup / documentation, so let's gather feedback and hopefully we have quite a bit of time to refine for 8.14.

LasseBlaauwbroek commented:
This will be extremely useful for me. Thanks for your efforts!

silene commented on coq-core.opam:
```diff
@@ -0,0 +1,34 @@
+  "ocaml"     { >= "4.05.0" }
+  "dune"      { >= "2.5.0"  }
+  "ocamlfind" { build }
+  "num"
```
Why was a dependency on `num` added?

Also, should the dependency on `dune` be declared `build` only? I do not understand why users should be forced to recompile Coq (and all the depending Coq libraries) whenever one updates Dune. (I am pretty sure I will never update Dune if I know it will cause the recompilation of all the Coq libraries.)

Zimmi48 replied:
> Also, should the dependency on `dune` be declared `build` only? I do not understand why users should be forced to recompile Coq (and all the depending Coq libraries) whenever one updates Dune. (I am pretty sure I will never update Dune if I know it will cause the recompilation of all the Coq libraries.)

IIRC Dune installs `dune-project` files which contains the Dune version. Dune tries as hard as it can to be backward-compatible but if you downgrade your Dune version after having installed some Dune package, then you break your installation. Anyway, this led to removing the `{ build }` annotation from every `dune` dependency in the opam-repository.

silene replied:
> Dune installs `dune-project` files which contains the Dune version.

Did you mean `dune-package`? Also, as far as I can tell, these files do not contain the version of Dune (but they contain the language version, which seems sensible to me).

> if you downgrade your Dune version after having installed some Dune package, then you break your installation

I have a hard time understanding why "downgrading" is a scenario worth making the life of every user painful. So yes, I guess I will never upgrade Dune, if it forces me to recompile every single package.

Zimmi48 replied:
Right. I don't know how frequently this happens. You should have a look at the relevant issues in the opam-repository for the context (I don't have the pointers, sorry).

ejgallego replied:
`num` seems like a bad rebase, thanks for catching it [I didn't do myself a super careful review until I fix the CI first]

`dune` cannot be `build` these days indeed, actually there is a very long discussion about that, as both dune-packages files and build actually are problematic, people is aware of it and trying to find a solution [I guess for dune 3.0] I suggest to discuss in the proper channel.

Note that if you enable compilation cache, and you upgrade dune, it should be expected that the recompilation time for packages building with dune to be close to zero. Not that this is a solution, but indeed changes the point of view a bit when build rules are effectively memoized.

ejgallego replied:
Resolved, that was a stale thing as we had to disable opam file generation.

Zimmi48 commented on Makefile.dune:
```diff
@@ -56,7 +56,8 @@ help-install:
 	@echo " Provided opam/dune packages are:"
 	@echo ""
-	@echo "  - coq: base Coq package, toplevel compilers, tools, stdlib, no GTK"
+	@echo "  - coq-core: base Coq package, toplevel compilers, tools, stdlib, no GTK"
```
```suggestion
	@echo "  - coq-core: base Coq package, toplevel compilers, tools, no stdlib, no GTK"
```

Zimmi48 commented on Makefile.install:
```diff
@@ -162,8 +162,9 @@ install-latex:
-	$(INSTALLLIB) META.coq $(FULLCOQLIB)/META
+install-meta: META.coq-core
+	$(INSTALLLIB) META.coq-core $(FULLCOQLIB)/META
+	rm -f $(FULLCOQLIB)/../coq-core && ln -s $(FULLCOQLIB) $(FULLCOQLIB)/../coq-core
```
I guess this is the compatibility line. Why is the `rm -f` needed?

Zimmi48 commented on coq-core.opam:
```diff
@@ -0,0 +1,34 @@
+
+depends: [
+  "ocaml"     { >= "4.05.0" }
+  "dune"      { >= "2.5.0"  }
```
In the Coq Call, it was said that this lower bound could be made even lower (for the pure OCaml coq-core package). But now that I think of it, this would require a separate `dune-project` file for `coq-core` and for `coq-stdlib`, since this is the file that specifies the minimal Dune version to use. That's a shame because I'm not even sure that this is a supported use case.

silene replied:
Due to #13437, I would on the contrary say that the bound should be put even higher, i.e., 2.8. Because it certainly does not work for me (unless something changed since I reported that bug).

Zimmi48 replied:
This bug should be closed now.

Note that the issue with 2.8 is that it is very recent and not yet packaged in Debian / Ubuntu, which would break @JasonGross's ppa, which is currently important for several projects' CI.

The plan laid out by @ejgallego during the call yesterday was that Dune would be required only to build the OCaml parts of Coq and there would be a choice of a coq_makefile-based or a Dune-based build for the standard library. If in addition, we could require different Dune lower bounds for the two parts (but I'm worried of this not being technically feasible), this plan would present the nice additional property of not forcing any recent Dune version upon our users.

ejgallego replied:
Yes we can uncouple the versions but that would require using different `dune-project` files for coq-core and coq-stdlib, so that'd be a further refinement.

Zimmi48 replied:
I think this worth it, but indeed, it can be done later. @JasonGross: you have no problem with Dune 2.5 becoming a requirement to build Coq, right?

JasonGross replied:
Ah, hold on, I should have looked into this more carefully earlier.  I've managed to package Dune 2.7 with OCaml 4.08.1 on all of the versions of Ubuntu that I care about, but I'm concerned that the 10x slowdown @ppedrot found at https://github.com/ocaml/ocaml/issues/7826#issuecomment-625342743 will hit fiat-crypto.  So it looks like I may need to move to 4.11.1 (can someone confirm that this version of OCaml does not suffer from this issue?), and I haven't yet packaged 4.11.1 for the various distributions I care about.  (And I may run into snags, as I did in packaging 4.08, such as OCaml being incompatible with the distributed version of binutils.)

Zimmi48 replied:
What about using Dune 2.7 compiled with OCaml 4.08.1 but staying on OCaml 4.07.1 for compiling Coq? Is this doable in your environment? If yes, that would be the safest solution.

ejgallego replied:
Indeed, OCaml 4.07 was release almost 3 years ago so it is safe start preparing the move from that version, as the GC parameter bug that was creating some problem has been finally solved, it is likely the recommended version becomes 4.12.0 soon.

We already discussed it, but it seems to me that using Debian packages to track Coq's master is gonna always be quite a bit of work, and will get worse in the future as document managers and other tooling are allowed to depend in even more packages. IMHO other solutions such as Nix / Docker / Opam are much preferable and tend to work out of the box. It makes little sense for the Coq team to have to worry about the complexities of Debian packaging as it is really a distribution issue, not related at all to Coq development itself so hard to justify to spend significant cycles on.

JasonGross replied:
[OCaml 4.11.1 does not even build on Groovy (Ubuntu 20.10), because many tests fail](https://launchpadlibrarian.net/525230922/buildlog_ubuntu-groovy-i386.ocaml_4.11.1-4.1~groovy~ppa1_BUILDING.txt.gz) with things like:
```
Running tests from 'tests/regression/pr3612' ...
 ... testing 'pr3612.ml' with 1 (native) => failed (The file /<<PKGBUILDDIR>>/testsuite/_ocamltest/tests/regression/pr3612/pr3612/ocamlopt.byte/ocamlopt.byte.output was expected to be empty because there is no reference file /<<PKGBUILDDIR>>/testsuite/tests/regression/pr3612/pr3612.compilers.reference but it is not:
========================================
/usr/bin/ld: pr3612.o: warning: relocation in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
========================================
```
What packages need to be upgraded or what needs to be done to fix this issue?

ejgallego replied:
No idea @JasonGross , I'm sorry, I wish I could be of more help but unfortunately I'm totally booked these days.

But that's a bit my point, very unfortunately, we don't have the resources in the Coq team these to be able to promise to help with there Coq-unrelated issues on a timely manner :( And on the other hand, our tooling / build systems needs a much needed refresh which we cannot delay more and will actually solve several dozen of issues with the current setup [such as rebases / bisects building properly, quick compilation, compilation caching, reduced maintenance, modularity, etc...]

I'd suggest tho you get in touch with Ralf Treinen and discuss.ocaml if you need Debian + OCaml help as I'm sure they will be much more helpful that we can [there is also OCaml discord and slack which is quite active]

ejgallego replied:
The build failure is surprising tho as the Debian packages for 4.11.1 work well; likely you need to update binutils, but this is a good example of why this crosses Coq's boundaries, updating binutils does impact almost _all_ packages in a Debian system so that is way out of our league.

Personally I'd abandon Debian packages as a CI platform and switch to Docker / Nix.

JasonGross replied:
Strangely enough, the build works fine in trusty, xenial, bionic, and focal (all of which are older than groovy) and hisrute (which is newer than groovy). Since groovy EOL's in July 2021, I can probably ignore this failure.

JasonGross replied:
I've made a post on [ocaml.discuss](https://discuss.ocaml.org/t/building-ocaml-4-11-1-on-ubuntu-20-10/7370/1), though it got hidden by a spam filter.  Is this the best platform for this?

ejgallego replied:
> I've made a post on [ocaml.discuss](https://discuss.ocaml.org/t/building-ocaml-4-11-1-on-ubuntu-20-10/7370/1), though it got hidden by a spam filter. Is this the best platform for this?

Ups ! I guess a moderator will take of it soon and approve your account, but indeed it is the best place to ask IMHO.

Zimmi48 commented on coq.opam:
```diff
@@ -20,11 +20,9 @@ homepage: "https://coq.inria.fr/"
+  "coq-core" {= version}
+  "coq-stdlib" {= version}
 ]
 build: [
```
Shouldn't you remove all build steps to make this a proper virtual package? And remove `dune` from the list of dependencies?

ejgallego replied:
Umm, good point; only point here would be to have to install the `dune-package` file, will ask.

ejgallego replied:
I am gonna leave it like this, as this could be useful for example to have the coq package run the test suite, but indeed we may want to further tweak; but indeed the dune subst, etc... is required to have proper pinning of Opam packages, so I think that due to this bug we should bump to 2.8 which while not ideal, it is convenient IMHO.

Zimmi48 reviewed (commented):
The Nix error is probably due to the test-suite being run *after* installation. This use case might be broken by the current state of the PR.

SkySkimmer commented:
Would it make sense to have a package for the prelude between coq-core and coq-stdlib?

silene commented:
I have been wondering as well but I concluded that it was not worth the effort. If some users have a need for the prelude, then they might also need some other files from the standard library, e.g., `Arith` or `List`. So, it feels like splitting just the prelude is not worth it. We might reconsider when we start splitting the standard library into larger chunks, e.g., `Reals`, `Fsets`.

Alizter commented:
Like @silene, I feel as though splitting prelude is really about splitting up the stdlib. It's not a given that everybody wants to use the same coq prelude for starters, and perhaps making the stdlib more modular so people can pick and chose parts of it is a more useful thing to do.

ejgallego commented:
Thanks to all, I think only the Nix issue do remain, and a few other nits that IMHO shouldn't be blocking.

> The Nix error is probably due to the test-suite being run after installation. This use case might be broken by the current state of the PR.

I think we may be doing something wrong with the way Nix installs OCaml libs, like the meta or our `..` hack not working. The test suite is just calling `ocamlfind query coq-core` and this is not working for Nix for some reason.

@Zimmi48 @vbgl any help here would be extremely welcome.

Zimmi48 commented on coq-core.opam:
```diff
@@ -0,0 +1,52 @@
+  "ounit2" {with-test}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
I thought this step was broken with Dune < 2.8...

ejgallego replied:
Yes, sorry, that was a fluke on my part [I temporary enable opam file generation to generate the opam files, then I tweak]

Zimmi48 commented on coq-doc.opam:
```diff
@@ -21,8 +21,7 @@ depends: [
 ]
 build: [
-# Disabled until Dune 2.8 is available
-# ["dune" "subst"] {pinned}
```
What justifies this change?

ejgallego replied:
Was a mistake, fixed thanks.

Zimmi48 commented on plugins/ltac/dune:
```diff
@@ -1,15 +1,19 @@
+ (libraries coq-core.plugins.ltac))
 
 (coq.pp (modules extratactics g_tactic  g_rewrite g_eqdecide g_auto g_obligations g_ltac profile_ltac_tactics coretactics g_class extraargs))
+
+; (deprecated_library_name
+;  (old_public_name coq.plugins.ltac)
+;  (new_public_name coq-core.plugins.ltac))
```
Any reason to leave this commented code in here?

ejgallego replied:
Resolved.

Zimmi48 commented:
One thing that was concluded in the Coq Call and not yet implemented here IINM is to improve the error message of `coqtop` when it cannot find the stdlib to indicate to users that they probably didn't install the correct package.

vbgl commented on lib/envars.ml:
```diff
@@ -205,6 +205,7 @@ let print_config ?(prefix_var_name="") f coq_src_subdirs =
   let open Printf in
   fprintf f "%sLOCAL=%s\n" prefix_var_name (if Coq_config.local then "1" else "0");
   fprintf f "%sCOQLIB=%s/\n" prefix_var_name (coqlib ());
+  fprintf f "%sCOQCORELIB=%s/\n" prefix_var_name (if Coq_config.local then coqlib () else coqlib () / "../coq-core/");
```
This is fishy: when the build is not local but the test-suite is run `coqlib ()` returns the path of the source. In this case, the sibling directory `coq-core` either does not exist or is unrelated.

ejgallego replied:
Umm, at least in the regular CI `coqlib` will return the installed path so indeed while hacky this should work, how do you call the test suite to get coqlib to point to the source in non-local build?

ejgallego replied:
Problem identified but was due to incorrect setting of COQLIB in coq makefile, actually the OSX job did also incorrectly set OCAMLPATH.

ejgallego commented:
> One thing that was concluded in the Coq Call and not yet implemented here IINM is to improve the error message of coqtop when it cannot find the stdlib to indicate to users that they probably didn't install the correct package.

Ups, I had written somewhere but I forgot, indeed will fix, thanks.

ejgallego commented:
Ok, thanks a lot for the help folks, this is looking more and more ready, CI should be fully green if I didn't do any silly mistake.

I think I have covered all the points raised, there is still a bit of hackery in some parts regarding paths which were already hacky but IMO the best approach there is to merge this, merge #13617 , and once that is done do the cleanup, and for example, remove all of our OCaml path handling in favor of ocamlfind.

ejgallego commented:
Note an important change since the last version [reflected in the changelog], we now require paths for `-prefix` to be absolute, after reviewing how paths are manipulated IMHO this is by far the best and safest solution, as opposed to actually checking lifting w.r.t. cd is done everywhere.

Zimmi48 commented on lib/envars.ml:
```diff
@@ -132,7 +132,8 @@ let guess_coqlib fail =
       else
-        fail "cannot guess a path for Coq libraries; please use -coqlib option")
+        fail "cannot guess a path for Coq libraries; please use -coqlib option \
+              or instruct Coq not to load the stdlib by using -boot -noinit parameters.")
```
This is not user-friendly enough. The message should contain an additional hint like "If you don't understand what this means, it probably means that you installed the wrong package (coq-core instead of coq)."

ejgallego replied:
New version.

Zimmi48 reviewed (commented):
Now it seems there's something off in the macOS build.

ejgallego commented:
> Now it seems there's something off in the macOS build.

I did a silly mistake setting ocamlpath, should work now.

Going to fix the comment, will push in a second.

Zimmi48 commented:
There still is a failure in the test-suite on macOS.

ejgallego commented:
> There still is a failure in the test-suite on macOS.

It turns out you need to perform `install-byte` too for the test-suite to be able to run from an installed Coq, Nix was already doing that. Crossing the fingers now.

Zimmi48 commented:
Will merge when CI passes.

Zimmi48 commented:
@coqbot merge now

Zimmi48 reviewed (approved):
@coqbot merge now

ejgallego commented:
Thanks to all! Let me know of any issue / question on Zulip.
